### DISCUSSION TITLE
손에 총기 또는 무기 붙이기

### DISCUSSION THREAD

**Discussion by smilejsu82** [naverz/zepeto-studio-kor#542] [5. Q&A]
![image](https://user-images.githubusercontent.com/50188795/160521963-3834d042-f98a-485b-8bb6-aaec354a7e44.png)
손에 총기 또는 무기를 부착하는 옳바른 방법은 무엇인가요?

스크립트로 동적으로 생성해 붙이고자 합니다 

![image](https://user-images.githubusercontent.com/50188795/160522090-1d22f086-d234-42f8-9124-5a91cc109355.png)

또한 자식 오브젝트를 찾는 방법을 제공하나요?
없다면 방법은 무엇인가요?
재귀적으로 찾을수 있게 직접 만들어 사용해야 하나요?
![image](https://user-images.githubusercontent.com/50188795/160523396-96dd9f05-3ab5-408a-9f21-0770baec7c28.png)

Hand_R 이 두개네요...
![image](https://user-images.githubusercontent.com/50188795/160524728-4d1c0d6f-3052-4e13-b849-e8d4c20dacbc.png)

일단 리소스를 수동으로 핸드 (Hand_R)에 자식으로 붙인뒤 애니메이션에 맞게 위치와 회전을 조절했습니다 
이후 Resources폴더에 넣고 동적으로 불러온뒤 프리팹 인스턴스를 생성하고 
Hand_R을 찾아 자식으로 붙였습니다 
이때 Hand_R이 두개여서 부모를 
var hips = this.zepetoCharacter.transform.Find("Zepeto Context/hips"); 과 같이 찾아왔습니다 
이후 Gun스크립트에 접근해 공격력을 셋팅해주었습니다 

이것이 옳바른 방법인지 알고 싶습니다 
(특히 핸드에 붙이는 과정)

https://user-images.githubusercontent.com/50188795/160525752-060a7ce1-b6d3-43a9-adda-e6efd147ebf9.mp4

**Comment by Zepeto-TechEvangelist** [comment]
안녕하세요. 
캐릭터의 특정 부위에 다른 오브젝트를 붙이는 방법은 여러가지가 있을 수 있지만
@smilejsu82 님이 구현하신 것과 마찬가지로 총 오브젝트를 손에 자식으로 붙인 후 위치값과 회전값을 조정해주는 것이 가장 간단한 방법입니다. 

아래 스크린샷 공식 예제 중 셀피 스틱을 잡는 기능 역시 같은 방법으로 구현하였습니다.

`예제 링크`: https://github.com/naverz/zepeto-multiplay-example/blob/main/Assets/ZepetoScripts/ScreenShotScripts/ScreenShotModeManager.ts

this.selfieStick = GameObject.Instantiate<GameObject>(this.selfieStickPrefab);
        this.localPlayer.character.GetComponentsInChildren<Transform>().forEach((characterObj) => {
            if(characterObj.name == this.rightHandBone) {
                this.selfieStick.transform.parent = characterObj;
                this.selfieStick.transform.localPosition = Vector3.zero;
                this.selfieStick.transform.localRotation = Quaternion.Euler(Vector3.zero);
                this.selfieStick.GetComponentInChildren<Renderer>().gameObject.layer = this.playerLayer;
            }
        });


원하는 본의 위치를 찾을 땐 위의 코드처럼 이름으로 찾을 수 있지만, 제페토 캐릭터들에 달린 애니메이터를 이용하여 찾을 수도 있습니다.

`참고 링크`: https://docs.unity3d.com/ScriptReference/Animator.GetBoneTransform.html

let targetCharacterAnimator = ZepetoPlayers.instance.LocalPlayer.zepetoPlayer.character.ZepetoAnimator;
let targetTransform = targetCharacterAnimator.GetBoneTransform(HumanBodyBones.RightHand);


한편, 총기를 잡을 때 Parent Constraint라는 컴포넌트를 이용하여 실제로 자식 오브젝트로 만들지 않더라도 마치 자식오브젝트인 것처럼 작동하게 만들 수도 있습니다.

https://docs.unity3d.com/kr/2020.3/Manual/class-ParentConstraint.html

**Reply by smilejsu82** [reply]
자세한 답변 정말 감사합니다 ~ 
참고해서 지금 코드를 좀 수정 해봐야 겠네요 
:)
